base/special/gamma.jl

```diff
@@ -630,12 +607,98 @@ function eta(z::Union{Float64,Complex{Float64}})
         return -zeta(z) * expm1(0.6931471805599453094172321214581765*δz)
     end
 end
-eta(x::Integer) = eta(Float64(x))
-eta(x::Real)    = oftype(float(x),eta(Float64(x)))
 
-"""
-    eta(x)
 
-Dirichlet eta function ``\\eta(s) = \\sum^\\infty_{n=1}(-1)^{n-1}/n^{s}``.
-"""
-eta(z::Complex) = oftype(float(z),eta(Complex128(z)))
+# Converting types that we can convert, and not ones we can not
+# If we really cared about single precision, we could make a faster
+# Float32 version by truncating the Stirling series at a smaller cutoff.
+# and if we really cared about half precision, we could make a faster
+# Float16 version, by using a precomputed table look-up.
+
+# Float16, and Float32 and their Complex equivalents can be converted to Float64
+# and results converted back. Similar for BitIntegers (eg Int32), but no converting back
+# Otherwise, we need to make things use their own `float` converting methods
+# and in those cases, we do not convert back either as we assume
+# they also implement there own versions of the functions
+
+
+ofpromotedtype(as::Tuple, c) = convert(promote_type(typeof.(as)...), c)
+ComplexOrRealUnion(TS...) = Union{(ComplexOrReal{T} for T in TS)...}
+
+const types_le_Float64 = (Float16, Float32, Float64, Base.BitInteger.types...)
+
+for T in types_le_Float64
+    @eval f64(x::Complex{$T}) = Complex128(x)
+    @eval f64(x::$T) = Float64(x)
+end
+
+
+for f in (:digamma, :trigamma, :zeta, :eta, :invdigamma)
+    @eval begin
+        $f(z::ComplexOrRealUnion(Base.BitInteger.types...)) = $f(f64(z))
```

Comment:
The problem is not accurasy (for real Ints at least), it is inconsistant return types.

Loss of accuracy does not occur until beyond ints 2^53 -- which I'm yet to find a case that it matters for any of these functions, as they tend to hit an asymptote long before then


We want `BitInteger` (`Int64` etc) to have `Float64` results from these functions (at least that is current behavior).

The main deal is we want `BigInt`s to have `BigFloat` results if possible,
and to throw method errors if not (rather than having `Float64` results).
Which is most of #17474

Eg right now 
```
julia> zeta(big"52")
1.0000000000000002

julia> zeta(big"52.0")
1.000000000000000222044605079804198399932009420465396423665432943893439236654051
```

We really want to avoid  reporting `BigFloat` results, that were calculated with `Float64`'s behind the scenes, and then converted. as in  the examples in https://github.com/JuliaLang/julia/pull/18584#issuecomment-248205809
I don't think that kinda thing ever happens for `BigInt`, right now.
Because right now most `BigIntegers` functions return `Float64s`
Even if there is more accurate `BigFloat` method available in MPFR